src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,390 @@
+        folders corresponding to it.
+
+    Returns:
+        Set[str]: Set of native dependency names. Note that we only return a set of strings here rather than Requirement
+        objects because the specific version of a native package is irrelevant.
+    """
+
+    def invert_downloaded_package_to_entry_map(
+        packages_dict: Dict[Requirement, List[str]]
+    ) -> Dict[str, Set[str]]:
+        """
+        Invert dictionary mapping packages to files/folders. We need this dictionary to be inverted because we first
+        discover files with native dependency extensions and then need to deduce the packages corresponding to these
+        files.
+
+        Args:
+            packages_dict (Dict[Requirement, List[str]]): Mapping between packages and a list of files or folders
+            corresponding to it.
+
+        Returns:
+            Dict[str, Set[str]]: The inverse mapping from a file or folder to the packages it corresponds to. Note that
+            it is unlikely a file corresponds to multiple packages (but we allow for the possibility). We only need
+            to return a set of strings here rather than Requirement objects because the specific version of a native
+            package is irrelevant.
+        """
+        record_entry_to_package_name_map: Dict[str, Set[str]] = {}
+        for requirement, record_entries in packages_dict.items():
+            for record in record_entries:
+                record_entry_to_package_name_map.setdefault(record, set()).add(
+                    requirement.name
+                )
+
+        return record_entry_to_package_name_map
+
+    native_libraries: Set[str] = set()
+    record_entries_to_package_map: Dict[
+        str, Set[str]
+    ] = invert_downloaded_package_to_entry_map(downloaded_packages_dict)
+
+    for native_extension in NATIVE_FILE_EXTENSIONS:
+        base_search_string: str = os.path.join(target, f"*{native_extension}")
+        recursive_search_string: str = os.path.join(
+            target, "**", f"*{native_extension}"
+        )
+
+        glob_output: List[str] = glob.glob(base_search_string) + glob.glob(
+            recursive_search_string, recursive=True
+        )
+        if glob_output and len(glob_output) > 0:
+            for path in glob_output:
+                relative_path = os.path.relpath(path, target)
+
+                # Fetch record entry (either base directory or a file name)
+                record_entry = os.path.split(relative_path)[0]
+                if (
+                    record_entry == ""
+                ):  # Implies the relative_path is a file name at the base directory
+                    record_entry = relative_path
+
+                # Check which packages own this record entry
+                if record_entry in record_entries_to_package_map:
+                    library_set = record_entries_to_package_map[record_entry]
+                    native_libraries.update(library_set)
```

Comment:
nit: is `library_set` the set of packages that depend on this native file? naming `package_set` makes it more readable

Reply:
updated